### PR TITLE
fix: Fixed bug with catboost and groups

### LINKED ISSUE #304: CatBoost Fails with Keyword 'groups'
Here's the code:

```python
settings = {
    "time_budget": 360,  
    "metric": 'ap',   
    "task": 'classification', 
    "log_file_name": f'{output_dir}/flaml1.log',  
    "seed": 7654321,    
    "log_training_metric": True,
    "groups": group_id,
    "estimator_list": ['catboost']  
}

automl.fit(X_train=X_train, y_train=y_train, **settings)
```

Here's the output:

```[flaml.automl: 11-23 14:03:00] {1485} INFO - Data split method: group
[flaml.automl: 11-23 14:03:00] {1489} INFO - Evaluation method: holdout
[flaml.automl: 11-23 14:03:05] {1540} INFO - Minimizing error metric: 1-ap
[flaml.automl: 11-23 14:03:05] {1577} INFO - List of ML learners in AutoML Run: ['catboost']
[flaml.automl: 11-23 14:03:05] {1826} INFO - iteration 0, current learner catboost

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-26-4e43c07e607c> in <module>
----> 1 automl.fit(X_train=X_train, y_train=y_train, **settings)

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/automl.py in fit(self, X_train, y_train, dataframe, label, metric, task, n_jobs, log_file_name, estimator_list, time_budget, max_iter, sample, ensemble, eval_method, log_type, model_history, split_ratio, n_splits, log_training_metric, mem_thres, pred_time_limit, train_time_limit, X_val, y_val, sample_weight_val, groups_val, groups, verbose, retrain_full, split_type, learner_selector, hpo_method, starting_points, seed, n_concurrent_trials, keep_search_state, early_stop, append_log, auto_augment, min_sample_size, use_ray, **fit_kwargs)
   1601             with training_log_writer(log_file_name, append_log) as save_helper:
   1602                 self._training_log = save_helper
-> 1603                 self._search()
   1604         else:
   1605             self._training_log = None

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/automl.py in _search(self)
   2117 
   2118         if not self._use_ray:
-> 2119             self._search_sequential()
   2120         else:
   2121             self._search_parallel()

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/automl.py in _search_sequential(self)
   1915                 time_budget_s=min(budget_left, self._state.train_time_limit),
   1916                 verbose=max(self.verbose - 3, 0),
-> 1917                 use_ray=False,
   1918             )
   1919             time_used = time.time() - start_run_time

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/tune/tune.py in run(training_function, config, low_cost_partial_config, cat_hp_cost, metric, mode, time_budget_s, points_to_evaluate, evaluated_rewards, prune_attr, min_resource, max_resource, reduction_factor, report_intermediate_result, search_alg, verbose, local_dir, num_samples, resources_per_trial, config_constraints, metric_constraints, max_failure, use_ray)
    402             if verbose:
    403                 logger.info(f"trial {num_trials} config: {trial_to_run.config}")
--> 404             result = training_function(trial_to_run.config)
    405             if result is not None:
    406                 if isinstance(result, dict):

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/automl.py in _compute_with_config_base(self, estimator, config_w_resource)
    241             self.learner_classes.get(estimator),
    242             self.log_training_metric,
--> 243             self.fit_kwargs,
    244         )
    245         result = {

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/ml.py in compute_estimator(X_train, y_train, X_val, y_val, weight_val, groups_val, budget, kf, config_dic, task, estimator_name, eval_method, eval_metric, best_val_loss, n_jobs, estimator_class, log_training_metric, fit_kwargs)
    435             budget=budget,
    436             log_training_metric=log_training_metric,
--> 437             fit_kwargs=fit_kwargs,
    438         )
    439     else:

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/ml.py in get_test_loss(config, estimator, X_train, y_train, X_test, y_test, weight_test, groups_test, eval_metric, obj, labels, budget, log_training_metric, fit_kwargs)
    258     #     fit_kwargs['X_val'] = X_test
    259     #     fit_kwargs['y_val'] = y_test
--> 260     estimator.fit(X_train, y_train, budget, **fit_kwargs)
    261     test_loss, metric_for_logging, pred_time, _ = _eval_estimator(
    262         config,

~/anaconda3/envs/python3/lib/python3.6/site-packages/flaml/model.py in fit(self, X_train, y_train, budget, **kwargs)
    939                 ),
    940                 callbacks=CatBoostEstimator._callbacks(start_time, deadline),
--> 941                 **kwargs,
    942             )
    943         else:

TypeError: fit() got an unexpected keyword argument 'groups'
```

### PR DESCRIPTION
**Pull request by dannycg1996**
## Why are these changes needed?

Currently an error is raised when we attempt to use a group-based split (such as `GroupKFold` or `StratifiedGroupKFold`) with CatBoost. This issue is caused by the `fit()` method on `CatBoostEstimator`, in particular this part:

```
            model.fit(
                X_tr,
                y_tr,
                cat_features=cat_features,
                eval_set=eval_set,
                callbacks=CatBoostEstimator._callbacks(
                    start_time, deadline, free_mem_ratio if use_best_model else None
                ),
                **kwargs,
            )
```

When `groups` is not None, kwargs contains the groups. CatBoost fit methods don't accept groups, which causes the issue.

As I understand it, groups in FLAML (and generally) should be used when splitting data, _not_ when fitting data. Basically, CatBoost does not need groups to be passed in, and likely never will.

I've been through the other FLAML estimators, and they all seem to remove groups from kwargs before calling `model.fit`. For example, in the `_fit()` method on the `BaseEstimator` class defined on `flaml/automl/model.py`, we have the following code:

```
        if "groups" in kwargs:
            kwargs = kwargs.copy()
            groups = kwargs.pop("groups")
            if self._task == "rank":
                kwargs["group"] = group_counts(groups)
```

This code removes `groups` from the kwargs dictionary, before `model.fit()` is called.

I've mimicked this 'removal of groups from kwargs' in my code. I've also updated the two tests of group-related splits to include CatBoost as an estimator. Before my changes, the addition of catboost causes these tests to fail, but after removing the groups from kwargs, they pass :)

I welcome any feedback!

## Related issue number

Closes #304 
## Checks

- [X] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks).
- [X] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by thinkall — COMMENTED** [review]
This is a trade-off in user experience. The current code raises an error when CatBoost is used with data containing groups, whereas your PR proposes to silently ignore the group information.

**Comment by dannycg1996** [comment]
> This is a trade-off in user experience. The current code raises an error when CatBoost is used with data containing groups, whereas your PR proposes to silently ignore the group information.

Hi @thinkall,

I'm confused by this. I would argue that my PR doesn't propose to silently ignore the group information - it is simply proposing to not pass groups into `.fit()`. This PR doesn't seek to treat CatBoost differently to any other estimator in FLAML - it's simply stripping `groups` out of the kwargs before passing the kwargs to `model.fit()`. I see no reason why we need to pass a `groups` argument into the `.fit()` method for CatBoost and CatBoost only.

To emphasis my point, I first created an example script which reproduces the reported CatBoost error:
```
from flaml import AutoML
from sklearn import datasets
import numpy as np

dic_data = datasets.load_iris(as_frame=True)  # numpy arrays
iris_data = dic_data["frame"]  # pandas dataframe data + target
rng = np.random.default_rng(42)
iris_data["cluster"] = rng.integers(
    low=0, high=5, size=iris_data.shape[0]
)
print(iris_data["cluster"])
automl = AutoML()
automl_settings = {
    "max_iter":2,
    "metric": 'accuracy',
    "task": 'classification',
    "log_file_name": "catboost_error.log",
    "log_type": "all",
    "estimator_list": ['catboost'],
    "eval_method": "cv",
    "split_type":"group",
    "groups": iris_data['cluster']
}
x_train = iris_data[["sepal length (cm)","sepal width (cm)", "petal length (cm)","petal width (cm)"]].to_numpy()
y_train = iris_data['target']
```
The error (as expected):
image

I then re-ran the script with `"estimator_list": ['lgbm']`. This initially runs without errors, as expected. However, if I go into `BaseEstimator._fit()`, on `flaml/automl/model.py and comment out this the kwarg-popping code mentioned above:
```
        if "groups" in kwargs:
            kwargs = kwargs.copy()
            groups = kwargs.pop("groups")
            if self._task == "rank":
                kwargs["group"] = group_counts(groups)
```
Then rerunning the AutoML code gives this error:

image

Likewise if I run the same code with `RandomForestEstimator` as my only estimator:
image

Basically as I understand it, `CatBoostEstimator` is the only FLAML estimator (that I use at least) which does not at any point call `BaseEstimator._fit()`. Therefore `CatBoostEstimator` is the only FLAML model which doesn't pop the `groups` keyword out of the kwargs before calling `model.fit()`. I've investigated this pretty thoroughly now, and I'm fairly confident that no other estimator tries to pass `groups` to `model.fit()`. I'm not sure why `CatBoostEstimator` should be the exception.

Apologies if I'm missing something obvious.
Anyway thanks for reading, I hope that makes sense - please let me know your thoughts!

**Review by thinkall — APPROVED** [review]
Thank you so much, @dannycg1996 !
